include/dlaf/permutations/general/impl.h

```diff
@@ -0,0 +1,201 @@
+#include "dlaf/sender/when_all_lift.h"
+#include "dlaf/types.h"
+#include "dlaf/util_matrix.h"
+
+#include "pika/parallel/algorithms/for_each.hpp"
+
+namespace dlaf::permutations::internal {
+
+// Applies the permutaton index `perm_arr` to a portion of the columns/rows(depends on coord) [1] of an
+// input submatrix [2] and saves the result into a subregion [3] of an output submatrix [4].
+//
+// Example column permutations with `perm_arr = [8, 2, 5]`:
+//
+//          2     5     8      out_begin
+//     ┌────────────────────┐    ┌─────┬─────────────┐
+//     │  in_offset      in │    │     │         out │
+//     │ │                  │    │     └─►┌─┬─┬─┐    │
+//     │◄┘ ┌─┐   ┌─┐   ┌─┐  │    │        │c│a│b│    │
+//     │   │a│   │b│   │c│  │    │    ┌──►│ │ │ │    │
+//     │   │ │ ┌►│ │   │ │◄─┼────┼──┐ │   └─┴─┴─┘    │
+//     │   └─┘ │ └─┘   └─┘  │    │  │ │      ▲       │
+//     │       │            │    │sz.rows()  │       │
+//     │      sz.rows()     │    │           └─      │
+//     │                    │    │          sz.cols()│
+//     └────────────────────┘    └───────────────────┘
+//
+// Example row permutations with `perm_arr = [3, 1]`:
+//
+//             ┌─── in_offset
+//             │                   out_begin
+//     ┌───────▼────────────┐    ┌──┬────────────────┐
+//     │                 in │    │  │            out │
+//     │       ┌─────┐      │    │  └►┌─────┐        │
+//   1 │       │  a  │      │    │    │  b  │        │
+//     │       └─────┘      │    │    ├─────┤ ◄─┐    │
+//     │                    │    │    │  a  │   │    │
+//     │       ┌─────┐      │    │    └──▲──┘   │    │
+//   3 │       │  b  │      │    │       │      │    │
+//     │       └──▲──┘      │    │ sz.cols()    │    │
+//     │          │         │    │          sz.rows()│
+//     └──────────┼─────────┘    └───────────────────┘
+//                │
+//             sz.cols()
+//
+//
+// [1]: The portion of each input column or row is defined by the interval [in_offset, in_offset +
+//      sz.col()) or the interval [in_offset, in_offset + sz.row()) respectively.
+// [2]: The input submatrix is defined by `begin_tiles`, `ld_tiles`, `distr` and `in_tiles`
+// [3]: The subregion is defined by `begin` and `sz`
+// [4]: The output submatrix is defined by `begin_tiles`, `ld_tiles`, `distr` and `out_tiles`
+//
+// Note: `in_tiles` should be `const T` but to avoid extra allocations necessary for unwrapping
+//       `std::vector<shared_future<matrix::Tile<const T, D>>>` it is left as non-const
+template <class T, Coord coord>
+void applyPermutations(GlobalElementIndex out_begin, GlobalElementSize sz, SizeType in_offset,
+                       const matrix::Distribution& distr, const SizeType* perm_arr,
+                       const std::vector<matrix::Tile<T, Device::CPU>>& in_tiles,
+                       const std::vector<matrix::Tile<T, Device::CPU>>& out_tiles) {
+  constexpr Coord ocrd = orthogonal(coord);
```

Comment:
`ocrd` = `Orthogonal CooRD`? Maybe `orth_coord` or something more explicit would be nice?